Api availability measurement

vamossagar12 opened the pull request:
This PR is for https://github.com/kubernetes/perf-tests/issues/1096 which created a new measurement to track the availability of api-server by probing the /healthz end point

mm4tt commented:
/ok-to-test

Thanks, @vamossagar12!

I'll try to take a look tomorrow. Adding Wojtek in case he has capacity to take a look today
/assign @wojtek-t

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,121 @@
+	ticker := time.NewTicker(500 * time.Millisecond)
+	stopAvailabilityCheck := make(chan bool)
+	apiServerClient := config.ClusterFramework.GetClientSets().GetClient().CoreV1().RESTClient()
+	numProbes := 0
```
Instead of having one huge Execute method, I would much rather prefer having a more organized approach with dedicate start() and gather() methods, those variables being measurement fields, etc.

As an example, PTAL at profile measurement, which is pretty similar conceptually to what you want to achieve:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/profile.go

vamossagar12 replied:
thanks @wojtek-t this was just an initial stab. That's why I created a draft PR to get comments and the direction i should be moving in.

vamossagar12 replied:
@wojtek-t thanks for pointing me to this example. As you said, it's pretty similar to what I want to achieve.
Couple of questions about interacting with api server. I see that there's explicit configs to interact with https and also add rbac roles to expose debug end point. Do you think we need all those features in this measurement?

wojtek-t replied:
> I  see that there's explicit configs to interact with https and also add rbac roles to expose debug end point. Do you think we need all those features in this measurement?

https is definitely needed. Re rbac - I can't remember from the top of my head how /health is handled - I think it's not needed, but it requires checking once you implement the measurement itself.

vamossagar12 replied:
ok. Another question that I have is, the profile measurement creates summaries during every run and appends them to a list. Do we want to have something similar? What I understood from the OP was that we want to print the statistics only at the end of the test?

wojtek-t replied:
We want to generate a single summary at the end. That said you need to keep track of the information you gather over time - so although slightly different, this will be fairly similar.

vamossagar12 replied:
yeah, i am already keeping track of some of those metrics. just need to surface them out at the end of the tests.

vamossagar12 replied:
hey i have made further changes to the measurements which are inline to the profile measurements and yet generating only a single summary at the end of gather phase.

oxddr commented:
/unassign

oxddr commented:
/uncc

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+	for _, host := range a.hosts {
+		// Get the api-server measurement data over SSH.
+		// Start by checking that the provider allows us to do so.
+		if !SSHToMasterSupported {
```
I don't think we should even consider SSH for this purpose.
Simply calling /healthz using client is what we want.

The problem is that using client you currently contact a random master. That said, maybe it's actually fine - if there is a single master down but load-balancer is not forwarding any requests to it, that sounds like a healthy cluster?

Thinking more about it, what we may consider is:
- healthiness at the cluster level
- healthiness of individual masters

wojtek-t replied:
[That said, I'm not sure if in CL2 we support passing "cluster-level IP" and "master IPs". Maybe we do and the first is via kubeconfig?]

vamossagar12 replied:
> I don't think we should even consider SSH for this purpose.
> Simply calling /healthz using client is what we want.
> 
> The problem is that using client you currently contact a random master. That said, maybe it's actually fine - if there is a single master down but load-balancer is not forwarding any requests to it, that sounds like a healthy cluster?
> 
> Thinking more about it, what we may consider is:
> 
> * healthiness at the cluster level
> * healthiness of individual masters

As per my understanding, If a single master is down and LB isn't forwarding any requests to it, then it can't be qualified as a healthy cluster. 

I think with the current implementation, for non-ssh mode, what we are considering is cluster level health. So, if a request returns, we count it as a success. 
It's only in the ssh case, where we consider hosts but in this case, if any one of the hosts reports an error, it is counted as a failure. So, that's an individual master level probe. 

So, you are saying we report both cluster/individual level modes? If we add individual level, then we will also need to track which masters failed how many times?

wojtek-t replied:
> As per my understanding, If a single master is down and LB isn't forwarding any requests to it, then it can't be qualified as a healthy cluster.

If there are multiple masters and one is down but not receiving any traffic, I claim that from user-perspective the cluster is healthy. And we care about user-perspective here.

> So, you are saying we report both cluster/individual level modes? If we add individual level, then we will also need to track which masters failed how many times?

yes - we need the same data for 1. cluster as a whole, 2. each master separately

vamossagar12 replied:
Understood. Do we want the health details for each master at every probe? If yes, then I will have to change the code - which I reused from the probes example that you had sent.
So, when SSH is not supported, then what we will have is only cluster level metrics.

It's only when ssh is supported that I can get individual level metrics. Here, currently from the re-used logic, if **any** of the master fails, the method returns a false for api server availability. Instead, I will have to ssh into each master, get the success/failure and report it back and update in a map where I will store metrics against each master. 

The only thing is, if we have too many masters then is there a chance of the probes overlapping with each other? Can that have a bearing on the metrics reported?

And lastly, is it a fair thing to say that if ssh isn't supported then all we get is cluster level metrics, and in the other case, both cluster level and individual master level metrics?

wojtek-t replied:
I think we should start with something and iterate, rather than trying to get a perfect solution from the very beginning:
- I'm fine initially that only return all metrics for ssh-supported providers [and that should be hidden behind flag anyway]
- re overlapping probes - that also sounds fine at least initially 

A clear TODOs should be added though to emhpasize the problems we're aware of.

vamossagar12 replied:
hey @wojtek-t , I have made the changes. I couldn't push them because formatting changes are failing for a recently pushed commit to probes.go. I have rectified it here:
https://github.com/kubernetes/perf-tests/pull/1190

Once merged, I will push my changes.

vamossagar12 replied:
Thanks for merging. I pushed the latest changes.

vamossagar12 replied:
hi @wojtek-t , did you get a chance to look at the latest changes?

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+	// if the previous probe was not a failure, then set the previous probe failure flag to true.
+	// This would be used by next probe
+	if !apiAvailabilityMetrics.prevProbeFailure {
+		apiAvailabilityMetrics.prevProbeFailure = true
```
Why do we need it? if consecutiveFailedProbes == 0 (or 1 at this poin in the code) that means that prevProbe wasn't a failure, otherwise it was).

vamossagar12 replied:
Well, I hadn't thought of doing it via the consecutiveFailedProbes counter. On second thoughts, when I re-looked at the code, the checks with a boolean flag seemed a bit more intuitive to understand that the previous probe was a failure than a counter based check. That's my personal opinion though and can change it if it seems complicated.

wojtek-t replied:
The less variables/fields we have the better.

If you want, you can introduce a method, like:
```
func (m *apiAvailabilitymetrics) previousProbeFailed() bool {
}
```

that will return what you want and use it where you want.
Let's not introduce unnecessary vars that you have to keep in sync.

vamossagar12 replied:
I made that change.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+	// if the probe succeeded, then increase the number of successful probes count by 1
+	apiAvailabilityMetrics.numSuccess++
+	// if this successful probe was received after failing probe(s)
+	if apiAvailabilityMetrics.prevProbeFailure {
```
This can be simplified a lot - see my comment above.

vamossagar12 replied:
Same as above.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+		return nil, err
+	}
+
+	probeDuration, err := util.GetInt(config.Params, "duration")
```
duration isn't the best name - it's rather frequency maybe?

wojtek-t replied:
We should also default to something sane, maybe 1s?

vamossagar12 replied:
True. Did both.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+		return a.start(config, SSHToMasterSupported, probeDuration)
+	case "gather":
+		a.stop()
+		availabilityPercentage := (a.apiAvailabilityMetrics.numSuccess / a.apiAvailabilityMetrics.numSuccess) * 100
```
Won't this always be 100?

vamossagar12 replied:
That was a bad miss, sorry. Updated.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+		return
+	}
+	close(a.stopCh)
+	a.wg.Wait()
```
That doesn't fully solve the problem - if the last period was the longest one, it won't be reported.

vamossagar12 replied:
Can you plz help me understand this scenario? Are you saying that when gather was called, there could have been another probe which would have made it the longest failing one and because I am closing the channel, it will miss that probe?
I have limited understanding of go routines but if the above scenario is correct, then does it mean that the channel stopping would override any currently executing probe? Should I wait before closing to check if any probes are ongoing?

wojtek-t replied:
This is the problem:
https://github.com/kubernetes/perf-tests/pull/1162/files#diff-769e85d1bdfa456cbbe72288f2ba2ab5R140-R142

If you have gazillions failure in a row at the end of the test, you will not include it in your summary.

vamossagar12 replied:
I added the comparision during the gather flow as well.

vamossagar12 commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+type apiAvailabilityOutput struct {
+	AvailabilityPercentage     int `json:"availabilityPercentage"`
```
Makes sense. I made the change.

vamossagar12 commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,202 @@
+
+type apiAvailabilityOutput struct {
+	AvailabilityPercentage     int `json:"availabilityPercentage"`
+	LongestUnavailableDuration int `json:"longestUnavailableDuration"`
```
I wasn't aware of the duration type in golang. Thay sounds more appropriate. I modified it.

wojtek-t commented:
@vamossagar12 - sorry for delay, I'm pretty overloaded. I will try to look into it today or tomorrow

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+)
+
+const (
+	apiAvailabilityName = "ApiAvailability"
```
s/ApiAvailability/APIAvailability/

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+func createApiAvailabilityMeasurement() measurement.Measurement {
```
s/Api/API/  (same everywhere else)

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+type apiAvailabilityMetrics struct {
+	numProbes                  int
```
numProbes seems unneded - it's numSuccess+numFailures

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+
+type apiAvailabilityMetrics struct {
+	numProbes                  int
+	numSuccess                 int
```
numSuccesses (plural, for consistency with failures)

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	wg                  sync.WaitGroup
+}
+
+func (a *apiAvailabilityMeasurement) updateMasterAvailabilityMetrics(c clientset.Interface, config *measurement.MeasurementConfig) {
```
This PR requires rebase - e.g. MeasurementConfig was renamed to Config.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	return nil, nil
+}
+
+func (a *apiAvailabilityMeasurement) updateAvailabilityMetrics(apiServerAvailable bool, metrics *apiAvailabilityMetrics) {
```
nit: move above the start method to keep this stuff together

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+		// SSH and check the health of the host
+		command := fmt.Sprintf("curl -s -k %slocalhost:%v/healthz", "https://", 443)
+		_, err := measurementutil.SSH(command, host+":22", "")
+		if err != nil {
```
Does it really do what you want? If the command fails with non-zero error code, is the err always returned?

vamossagar12 replied:
Do you mean that err won't capture failure of command with non-zero error code?

wojtek-t replied:
Yes - I think so (which is kind-of by-design):
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/util/ssh.go#L65
https://github.com/kubernetes/perf-tests/blob/2eeb004780c3f0ec99d0e03e6e44261e03017d6f/clusterloader2/vendor/k8s.io/kubernetes/pkg/ssh/ssh.go#L217

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+func (a *apiAvailabilityMeasurement) start(config *measurement.MeasurementConfig, SSHToMasterSupported bool, probeDuration int) ([]measurement.Summary, error) {
+	a.hosts = config.ClusterFramework.GetClusterConfig().MasterIPs
+	if len(a.hosts) < 1 {
+		return nil, goerrors.Errorf("Api Server measurement can't execute due to no MasterIps")
```
APIAvailability measurement ...

s/MasterIps/MasterIPs/

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	a.isRunning = true
+	a.stopCh = make(chan struct{})
+	a.wg.Add(1)
+	go func() {
```
nit: add an empty line before this go func - it's separating intialization from actual execution

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+				if SSHToMasterSupported {
+					a.updateMasterAvailabilityMetrics(k8sClient, config)
+				}
+				return
```
that seems incorrect - we don't want to return after the first probe, right?

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+// If the consecutive failed probes counter value > 0, then it means we are in the middle of an unavailability period
```
nit: please make this line shorter

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+// Check if the recently concluded consecutive failed number of probes is
+// higher than the max consecutive failed number of probes so far
+// if yes, then Update max consecutive failed probes
+func (a *apiAvailabilityMetrics) mayBeUpdateMaxConsecutiveFailures() {
```
updateMaxConsecutiveFailuresIfNeeded

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+func (a *apiAvailabilityMetrics) updateFailureMetrics() {
```
Can you maybe please move apiAvailabilityMetrics type and its methods to a separate file?

vamossagar12 replied:
Sure, would it make sense to move them to the utils package?

wojtek-t replied:
They are pretty specific to this measurement, so I don't think utils package is the best.
At some point, we should probably refactor each measurement to its own subpackage, but that's independent from this PR for sure.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+func (a *apiAvailabilityMetrics) updateFailureMetrics() {
+	// increment total number of failed probes count, current number of failed probes by 1
```
remove this comment - don't describe what the code is doing - only describe why if that isn't obvious (in this case it is, so no comment is needed)

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+func (a *apiAvailabilityMetrics) updateSuccessMetrics() {
+	// if the probe succeeded, then increase the number of successful probes count by 1
```
same here and below

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+// Execute starts the api-server healthz probe end point from start action and
+// collects availability metrics in gather.
+func (a *apiAvailabilityMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+
```
nit: remove this empty line

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+func (a *apiAvailabilityMeasurement) createClusterAvailabilitySummary(probeFrequency int, summaries []measurement.Summary) []measurement.Summary {
+	a.clusterLevelMetrics.mayBeUpdateMaxConsecutiveFailures()
```
let's not do this here - instead let's make it a defer in stop() function (or not a defer - just done after wg.Wait()

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	a.clusterLevelMetrics.mayBeUpdateMaxConsecutiveFailures()
+	clusterAvailabilityMetricsOutput := a.buildApiAvailabilityMetricsOutput(probeFrequency, a.clusterLevelMetrics, true, "")
+	content, err := util.PrettyPrintJSON(clusterAvailabilityMetricsOutput)
+	if err == nil {
```
don't ignore the error - if error != nil this function should return an error

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	for _, host := range a.hosts {
+		a.hostLevelMetrics[host].mayBeUpdateMaxConsecutiveFailures()
+		hostAvailabilityMetricsOutput := a.buildApiAvailabilityMetricsOutput(probeFrequency, a.hostLevelMetrics[host], false, host)
+		content, err := util.PrettyPrintJSON(hostAvailabilityMetricsOutput)
```
don't ignore the error - if error != nil this function should return an error

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+
+func (a *apiAvailabilityMeasurement) createMasterAvailabilitySummary(probeFrequency int, summaries []measurement.Summary) []measurement.Summary {
+	for _, host := range a.hosts {
+		a.hostLevelMetrics[host].mayBeUpdateMaxConsecutiveFailures()
```
Same here - this should be done in stop() function

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+func (a *apiAvailabilityMeasurement) start(config *measurement.MeasurementConfig, SSHToMasterSupported bool, probeDuration int) ([]measurement.Summary, error) {
```
start shouldn't return measurement.Summary - simply return just error

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	return summaries
+}
+
+func (a *apiAvailabilityMeasurement) buildApiAvailabilityMetricsOutput(probeFrequency int, metrics *apiAvailabilityMetrics, clusterMetrics bool, hostIp string) *apiAvailabilityOutput {
```
this shouldn't be apiAvailabilityMeasurement method (a isn't used anywhere).

Instead, make this a method of apiAvailabilityMetrics

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	// Gather availability metrics
+	availabilityPercentage := (float32(metrics.numSuccess) / float32(metrics.numProbes)) * 100
+	longestUnavailableDuration := metrics.maxConsecutiveFailedProbes * probeFrequency
+	availabilityMetrics := &apiAvailabilityOutput{AvailabilityPercentage: availabilityPercentage, LongestUnavailableDuration: time.Duration(longestUnavailableDuration), ClusterMetrics: clusterMetrics}
```
nit: please don't try to put it into single line

vamossagar12 commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,248 @@
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/util"
+	"sync"
+	"time"
```
done

wojtek-t replied:
not done

vamossagar12 replied:
yeah sorry about that. i haven't pushed the changes yet

vamossagar12 commented:
Thanks @wojtek-t I have more or less implemented the changes you suggested. Just had a couple of questions post which I can push the changes.

vamossagar12 commented:
@wojtek-t i made the changes but not able to push them. Keep getting this error after rebasing with the latest from master:

.git/hooks/pre-push: line 11: ./verify/verify-boilerplate.sh: No such file or directory

I see some changes have been made recently in the /verify folder but not able to figure out what the problem is.

wojtek-t commented:
@vamossagar12 - the verify-boilerplace.sh script still exists and wasn't touched:
https://github.com/kubernetes/perf-tests/blob/master/verify/verify-boilerplate.sh

BTW - the hooks must be something in your local environment - I don't have them set up.

vamossagar12 commented:
> @vamossagar12 - the verify-boilerplace.sh script still exists and wasn't touched:
> https://github.com/kubernetes/perf-tests/blob/master/verify/verify-boilerplate.sh
> 
> BTW - the hooks must be something in your local environment - I don't have them set up.

Well, actually I set the hooks up since they were mentioned in the README of perf-tests. If it isn't needed, I can remove it. Would it have any implications though?

wojtek-t commented:
> Well, actually I set the hooks up since they were mentioned in the README of perf-tests. If it isn't needed, I can remove it. Would it have any implications though?

We have presubmits anyway - so these should be caught by them anyway.

vamossagar12 commented:
> > Well, actually I set the hooks up since they were mentioned in the README of perf-tests. If it isn't needed, I can remove it. Would it have any implications though?
> 
> We have presubmits anyway - so these should be caught by them anyway.

Alright. Removed the local hooks folder and was able to push now.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,201 @@
+		// SSH and check the health of the host
+		command := fmt.Sprintf("curl -s -k %slocalhost:%v/healthz", "https://", 443)
+		sshResult, err := measurementutil.SSH(command, host+":22", "")
+		if err != nil || sshResult.Code != 0 {
```
What does curl return, if the server return say 500?

vamossagar12 replied:
I think it would be part of the stdout. So, might want to check for the status. I am not totally sure of the behaviour of the ssh session which executes the command. Is there a way to test?

wojtek-t replied:
You can start locally a server that will always respond 500 and call this code against it.

vamossagar12 replied:
I was trying to write a test for this with the same ssh code. I have docker for desktop running on my local and wanted to ssh into a locally running kube-api server. The master IP that I am passing is kubernetes.docker.internal. I see that in the configs, there is an option to pass a MasterIP. But, on a locally running kube cluster, masterIP is local, right?

Note that, I am able to run the /healthz end points by setting up a clientset with the same kube config file and the below example you provided works.

vamossagar12 replied:
@wojtek-t I was little jammed with work this week but tried some more things today. I set up a docker container running a Flask application inside and deployed it through kubernetes. i also set up a mechanism to ssh into the container. 

Then why I ran it through the test case which invokes this logic, i found that any error in web app is returned in sshResult.stderr. I tried with 404 and got a big html like string (perhaps flask returns that). Note that sshResult.stdout is empty.

In the happy scenario, the sshResult.stdout has the response and sshResult.stderr is empty. 

So, to answer the original question, can I also make this assumption that the presence of stderr indicates the /healthz end point failed and vice versa in the other case? Should I also look to parse the stdout result?

PS Unfortunately, I still couldn't find a way to ssh into the kube-apiserver container. Otherwise I could have validated directly.

vamossagar12 replied:
hey @wojtek-t i am presuming you're busy... Just reaching out to ask did you get a chance to look at the comments?

wojtek-t replied:
Sorry for delay.
It would be better to check it directly, but I don't want to block the whole PR on this thing.
So if you could open an issue to check the semantics of ssh and document it better and here in the code add a TODO (linked to this issue) to validate if the logic of relying on stderr non-empty is ok, I'm fine with that (resolving it should be a blocker for enabling the measurement though, but not necessary for merging this PR).

vamossagar12 replied:
Thanks @wojtek-t . As suggested, created a new issue to understand ssh client semantics:
https://github.com/kubernetes/perf-tests/issues/1343 and added a TODO here.
Just one thing, you asked me to document it you meant in the newly created issue?

vamossagar12 replied:
hey @wojtek-t I have the made the changes to this PR based upon my findings mentioned here:
https://github.com/kubernetes/perf-tests/issues/1343

Plz review whenever you have the time :)

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,201 @@
+
+func (a *apiAvailabilityMeasurement) updateClusterAvailabilityMetrics(c clientset.Interface) {
+	// Check the availability of the cluster by issuing a REST call to /healthz end point
+	_, err := c.CoreV1().RESTClient().Get().AbsPath("/healthz").DoRaw()
```
Don't we need to parse the response to get the returned code?

vamossagar12 replied:
Yeah that's needed. Not totally sure what the /healthz end point out put be though. Any way to find it?

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,201 @@
+	a.clusterLevelMetrics.updateMaxConsecutiveFailuresIfNeeded()
+	clusterAvailabilityMetricsOutput := a.clusterLevelMetrics.buildAPIAvailabilityMetricsOutput(probeFrequency, true, "")
+	content, err := util.PrettyPrintJSON(clusterAvailabilityMetricsOutput)
+	if err == nil {
```
if err != nil {
  return err
}
// the rest of logic

[less nesting and easier to follow]

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,201 @@
+		a.hostLevelMetrics[host].updateMaxConsecutiveFailuresIfNeeded()
+		hostAvailabilityMetricsOutput := a.hostLevelMetrics[host].buildAPIAvailabilityMetricsOutput(probeFrequency,false, host)
+		content, err := util.PrettyPrintJSON(hostAvailabilityMetricsOutput)
+		if err == nil {
```
if err != nil {
   ...
   continue
}
// rest of logic

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,201 @@
+			summary := measurement.CreateSummary(apiAvailabilityName, "json", content)
+			a.summaries = append(a.summaries, summary)
+		} else{
+			createMasterAvailabilitySummaryErrors.Append(createMasterAvailabilitySummaryErrors, err)
```
I think you only want to append err

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_metrics.go:
```diff
@@ -0,0 +1,86 @@
+	LongestUnavailableDuration time.Duration `json:"longestUnavailableDuration"`
+}
+
+// number of consecutive failed probes is > 0 signifies previous probe was a failure
```
// previousProbeFailed returns whether the last probe failed healthcheck.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_metrics.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// number of consecutive failed probes is > 0 signifies previous probe was a failure
+func (a *apiAvailabilityMetrics) previousProbeFailure() bool {
```
previousProbeFailed

wojtek-t replied:
Actually - let's remove this method completely - see my comment below.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_metrics.go:
```diff
@@ -0,0 +1,86 @@
+	return a.consecutiveFailedProbes > 0
+}
+
+// Check if the recently concluded consecutive failed number of probes is
```
follow the gostyle: comments should start with method name.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_metrics.go:
```diff
@@ -0,0 +1,86 @@
+
+func (a *apiAvailabilityMetrics) updateSuccessMetrics() {
+	a.numSuccesses++
+	if a.previousProbeFailure() {
```
let's remove this check and do that unconditionally - it's not needed.

if the previous attempt was success, consecutiveFailedProbes=0, but maxing with 0 and setting it to 0 don't hurt

vamossagar12 commented:
> Also please squash commits

Done. Had a couple of questions.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,207 @@
+
+func (a *apiAvailabilityMeasurement) updateClusterAvailabilityMetrics(c clientset.Interface) {
+	// Check the availability of the cluster by issuing a REST call to /healthz end point
+	_, err := c.CoreV1().RESTClient().Get().AbsPath("/healthz").DoRaw(context.TODO())
```
I think what you want is this one:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/testing/testserver.go#L174-L179

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+func (a *apiAvailabilityMeasurement) createClusterAvailabilitySummary(probeFrequency int) error {
+	a.clusterLevelMetrics.updateMaxConsecutiveFailuresIfNeeded()
```
Instead of doing this here, let's:
- remove it from here (and createMasterAvailabilitySummary)
- let's explicitly add it `updateFailureMetrics`, i.e.;

```
func (a *apiAvailabilityMetrics) updateFailureMetrics() {
	a.numFailures++
	a.consecutiveFailedProbes++
        updateMaxConsecutiveFailuresIfNeeded()
}
```

With that - you don't need to call it anywhere elese.

vamossagar12 replied:
Done. That makes sense.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,209 @@
+	createMasterAvailabilitySummaryErrors := errors.NewErrorList()
+	for _, host := range a.hosts {
+		a.hostLevelMetrics[host].updateMaxConsecutiveFailuresIfNeeded()
+		hostAvailabilityMetricsOutput := a.hostLevelMetrics[host].buildAPIAvailabilityMetricsOutput(probeFrequency,false, host)
```
nit: space before 'false"

vamossagar12 replied:
Done.

vamossagar12 commented:
> Sorry for delay - I was on vacation for last 2.5 weeks.
> 
> I added some comments, but i'm mostly fine with what is now. Please change the PR out of Draft and let's proceed with it.

No worries :) I have made the changes and also removed the PR out of draft.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,207 @@
+		content, err := util.PrettyPrintJSON(hostAvailabilityMetricsOutput)
+		if err != nil {
+			createMasterAvailabilitySummaryErrors.Append(err)
+		} else{
```
Please fix gofmt:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1162/pull-perf-tests-verify-lint/1285548209413820419/build-log.txt

[both here and in the second file]

vamossagar12 replied:
I ran gofmt -s for these 2 files but this error still got thrown. Is that the right command?

wojtek-t replied:
gofmt -w -s

vamossagar12 replied:
thank you.. ran the command and pushed again.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,207 @@
+func (a *apiAvailabilityMeasurement) start(config *measurement.Config, SSHToMasterSupported bool, probeDuration int) error {
+	a.hosts = config.ClusterFramework.GetClusterConfig().MasterIPs
+	if len(a.hosts) < 1 {
+		return goerrors.Errorf("API Server measurement can't execute due to no MasterIPs")
```
nit: APIAvailability measurement can't start due to lack of master IPs

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,207 @@
+	a.wg.Wait()
+
+	errList := errors.NewErrorList()
+	err := a.createClusterAvailabilitySummary(probeFrequency)
```
Sorry - I should have caught that earlier.

Actually I would really prefer to have a single summary rather than N summaries for every master and cluster separately (see my comment below).

wojtek-t replied:
This isn't yet addressed.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_metrics.go:
```diff
@@ -0,0 +1,81 @@
+	consecutiveFailedProbes    int
+}
+
+type apiAvailabilityOutput struct {
```
Actually, I would prefer the output to be:

```
type apiAvailabilitySummary struct {
  IP                                           string
  AvailabilityPercentage           float32
  LongestUnavailableDuration time.Duration
}

type apiAvailabilityOutput struct {
  ClusterMetrics apiAvailabilitySummary
  MasterMetrics []apiAvailabilitySummary
}
```

vamossagar12 replied:
hi @wojtek-t , slight confusion here. The final output for MasterMetrics is an array here. If you see the output of the measurement that would be returned, all outputs would be consolidated into 1. 

https://github.com/kubernetes/perf-tests/pull/1162/files#diff-769e85d1bdfa456cbbe72288f2ba2ab5R54-R55

Do you think this isn't similar to what you prefer?

wojtek-t replied:
I care about 'availability' summary as a whole. Yes - I want a single summary, instead of N separate ones.
Alternatively, we may consider two measurements: one for the cluster, second for all masters.

vamossagar12 replied:
I see. I might want to make some tweaks to the apiAvailabilityMeasurement struct in that case. Currently it looks like this:

```
type apiAvailabilityMeasurement struct {
	isRunning           bool
	stopCh              chan struct{}
	hosts               []string
	summaries           []measurement.Summary
	clusterLevelMetrics *apiAvailabilityMetrics
	hostLevelMetrics    map[string]*apiAvailabilityMetrics
	wg                  sync.WaitGroup
}
```
So, if I understand correctly, instead of an array of summaries, you want 1 cluster level summary and array of MasterLevel summaries in a single object. I would also need to remove the clusterLevelMetrics and hostLevelMetrics from the above struct.

Regarding 2 separate measurements, do we expose them as separate measurements- which seems slightly confusing to me. With a unified measurement, a user would be able to get all availability metrics in one summary.

wojtek-t replied:
Sorry - I meant "two summaries" - one for the cluster, second for all masters.
But, I think having a single summary is cleaner.

vamossagar12 replied:
ok. so in that case, i will make the code changes to have one summary in the format you explained above.

vamossagar12 replied:
@wojtek-t I have made the code changes to expose only 1 summary for both ClusterLevel and MasterMetrics. Plz review when you get the chance..

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_metrics.go:
```diff
@@ -0,0 +1,81 @@
+func (a *apiAvailabilityMetrics) updateSuccessMetrics() {
+	a.numSuccesses++
+	if a.consecutiveFailedProbes > 0 {
+		a.updateMaxConsecutiveFailuresIfNeeded()
```
this is no longer needed - you already checked that in updateFailureMetrics and here you don't bump numConsecutiveFailures

wojtek-t replied:
So basically, simply unconditionally set consecutiveFailedProbes = 0

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,211 @@
+			a.updateAvailabilityMetrics(false, a.hostLevelMetrics[host])
+		} else {
+			a.updateAvailabilityMetrics(true, a.hostLevelMetrics[host])
+		}
```
Let's switch to:
```
availability := err != nil && sshResult.Code != 0
a.udpateAvailabilityMetrics(availability, a.hostLevelMetrics[host])
```

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,211 @@
+		a.updateAvailabilityMetrics(true, a.clusterLevelMetrics)
+	} else {
+		a.updateAvailabilityMetrics(false, a.clusterLevelMetrics)
+	}
```
a.updateAvailabilitymetrics(status == 200, a.clusterLevelMetrics)

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,211 @@
+// Execute starts the api-server healthz probe end point from start action and
+// collects availability metrics in gather.
+func (a *apiAvailabilityMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+
```
nit: remove this empty line

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return nil, err
+	}
+
```
nit: remove this empty line too

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,211 @@
+	content, err := util.PrettyPrintJSON(clusterAvailabilityMetricsOutput)
+	if err != nil {
+		return err
+	} else {
```
```
if err != nil {
  return err
}
a.summaries = append...
return nil
```

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,185 @@
+	summaries             []measurement.Summary
+	clusterLevelMetrics   *apiAvailabilityMetrics
+	hostLevelMetrics      map[string]*apiAvailabilityMetrics
+	apiAvailabilityOutput *apiAvailabilityOutput
```
let's remove this field - this concenptually isn't part of the measurement. See also my comments below.

wojtek-t commented on clusterloader2/pkg/measurement/common/api_availability_measurement.go:
```diff
@@ -0,0 +1,185 @@
+	close(a.stopCh)
+	a.wg.Wait()
+
+	a.createClusterAvailabilitySummary(probeFrequency)
```
Given apiAvailabilityOutput shouldn't be a measurement-level field, let's simply do:
```
output := apiAvailabilityOutput{}
output.createClusterSummary(a.clusterLevelMetrics, probeFrequency)
if SSHToMasterSupported {
  output.createMastersSummary(a.hostLevelMetrics, probeFrequency)
}
```

[which also means removing the createSummary methods above and making the output methods]

vamossagar12 commented:
Thanks @wojtek-t , I made those changes as well.

wojtek-t commented:
/lgtm
/approve
